### ISSUE TITLE
Yet more misparsing: invalid parentheses in call statements and assignments

### ISSUE THREAD

**Issue by fstirlitz** [fstirlitz/luaparse#24]
Each of the lines that follow is a syntax error, yet luaparse does not recognise these as such:

``` lua
(a.b) = 0
(a) = 0
a:b() = 0
a() = 0
a.b:c() = 0
a[b]() = 0
(0) = 0
a, b() = 0, 0
```

Note, however, that the following are not (syntactical, at least) errors:

``` lua
a().b = 1
({})[b] = 1
a""[b] = 1
a{}[b] = 1
({{}})[a][b] = 1
(a).b = 1
(1).a = 2 -- runtime error, unless you use debug.setmetatable
```

Relevant portions of the Lua 5.2 manual (essentially identical in Lua 5.0 and 5.1, and to 5.3-work2): [§3.2 "Variables"](http://lua.org/manual/5.2/manual.html#3.2), [§3.3.3 "Assignment"](http://lua.org/manual/5.2/manual.html#3.3.3), [§9 "The Complete Syntax of Lua"](http://lua.org/manual/5.2/manual.html#9)

**Comment by StarveTheEgo** [comment]
First time using GitHub for some discussions. Found out that if you type, for example,

local test_obj = {a = b c = e} -- @note: missed comma here
It won`t find any error here

**Comment by fstirlitz** [comment]
Should also add:
```lua
(a());
(a"");
(a{});
```

I've got some WIP code which should fix this once and for all.

**Comment by berbun** [comment]
Hello.
I'd like to append two more (unterminated single-line strings):
`local s="`
`local s="`
(unterminated multi-line string seems to raise an exception good.)

**Comment by fstirlitz** [comment]
I was pretty sure this was tested already, but apparently not strictly enough.

But either way, this should be filed separately. I should probably edit the title to reflect this.

**Comment by berbun** [comment]
Here's an interesting case I run into:
`f--[[]]('x')--[[]]['y']--[[]]=9--`
Results in:
> SyntaxError: [1:0] unexpected identifier 'f' near '<eof>'

But. Adding a comment before everything:
`--[[]]f--[[]]('x')--[[]]['y']--[[]]=9--`
Then, it works. 😕 🤔 😕

**Comment by fstirlitz** [comment]
I cannot reproduce this, and it should have been filed separately anyway.

**Comment by berbun** [comment]
> I cannot reproduce this, and it should have been filed separately anyway.

Sorry, I forgot I made some changes on the luaparse. Apparently have done it wrong.
That was my bad.
Move on.

**Comment by fstirlitz** [comment]
Finally definitively fixed in 6c00a3a6193f6baff802c03a4bf6fd0c63c01eba. The particular parsing mechanics (and the resulting error messages) may still change in the future, but at least all test cases fail as expected now.

**Comment by arnoson** [comment]
How can I distinguish whether an expressions was wrapped in parens?
Right now these two produce the same AST
```lua
local a, b = test()
```
```lua
local a, b = (test())
```
